chain/consensus/actors/atomic-replace/replace.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/filecoin-project/specs-actors/v7/actors/runtime"
+	cid "github.com/ipfs/go-cid"
+	xerrors "golang.org/x/xerrors"
+)
+
+//go:generate go run ./gen/gen.go
+
+// example "Replace" actor that atomically replaces the cid from one owner
+// to the other.
+
+var _ runtime.VMActor = ReplaceActor{}
+var _ atomic.LockableActor = ReplaceActor{}
+
+// ReplaceState determines the actor state.
+// FIXME: We are using a non-efficient locking strategy for now
+// where the whole map is locked for an atomic execution.
+// We could use a more fine-grained approach, but this actor
+// has illustrative purposes for now. Consider improving it in
+// future iterations.
+type ReplaceState struct {
+	Owners *atomic.LockedState
+}
+
+type Owners struct {
+	M map[string]cid.Cid
+}
+
+// Merge implement the merge strategy to follow for our lockable state
+// in the actor.
+func (o *Owners) Merge(other atomic.LockableState) error {
+	tt, ok := other.(*Owners)
+	if !ok {
+		return xerrors.Errorf("type of LockableState not Owners")
+	}
+
+	for k, v := range tt.M {
+		_, ok := o.M[k]
+		if ok {
+			return xerrors.Errorf("merge conflict. key for owner already set")
+		}
+		o.M[k] = v
+	}
+	return nil
+
+}
+
+func ConstructState() (*ReplaceState, error) {
+	owners, err := atomic.WrapLockableState(&Owners{M: map[string]cid.Cid{}})
+	if err != nil {
+		return nil, err
+	}
+	return &ReplaceState{Owners: owners}, nil
+}
+
+const (
+	MethodReplace = 6
+	MethodOwn     = 7
+)
+
+type ReplaceActor struct{}
+
+func (a ReplaceActor) Exports() []interface{} {
+	return []interface{}{
+		builtin.MethodConstructor: a.Constructor,
+		atomic.MethodLock:         a.Lock,
+		atomic.MethodMerge:        a.Merge,
+		atomic.MethodAbort:        a.Abort,
+		atomic.MethodUnlock:       a.Unlock,
+		MethodReplace:             a.Replace,
+		MethodOwn:                 a.Own,
+	}
+}
+
+func (a ReplaceActor) Code() cid.Cid {
+	return actor.ReplaceActorCodeID
+}
+
+func (a ReplaceActor) IsSingleton() bool {
+	return false
+}
+
+func (a ReplaceActor) State() cbor.Er {
+	return new(ReplaceState)
+}
+
+func (a ReplaceActor) Constructor(rt runtime.Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
+	rt.ValidateImmediateCallerType(builtin.InitActorCodeID)
+	st, err := ConstructState()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error computing initial state")
+	rt.StateCreate(st)
+	return nil
+}
+
+type OwnParams struct {
+	Seed string
+}
+
+func (a ReplaceActor) Own(rt runtime.Runtime, params *OwnParams) *abi.EmptyValue {
+	rt.ValidateImmediateCallerAcceptAny()
+
+	var st ReplaceState
+	rt.StateTransaction(&st, func() {
+		ValidateLockedState(rt, &st)
+		own, err := st.UnwrapOwners()
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error unwrapping lockable state")
+		_, ok := own.M[rt.Caller().String()]
+		if ok {
+			rt.Abortf(exitcode.ErrIllegalState, "address already owning something")
+		}
+		own.M[rt.Caller().String()], err = abi.CidBuilder.Sum([]byte(params.Seed))
```

Comment:
what if params is nil?

Reply:
I don't think params can ever be nil. Actor methods receive their params after unmarshalling a stream. It would fail somewhere before this can be nil. That being said, I agree that having a sanity-check in critical actors (which is not currently the case) may not be a bad idea.